File: src/commands/task.ts
// `specd task <slug> <id> --status <s> [--evidence|--reason]` — the evidence gate (SPEC §5.2, §3.1).
import { requireSpecdRoot } from "../core/paths.js";
import { artifactPath, loadSpec } from "../core/specFiles.js";
import { withSpecLock } from "../core/lock.js";
import { atomicWrite, readOrDefault } from "../core/io.js";
import { usageError, gateError, notFoundError } from "../core/exit.js";
import { nowIso, saveState, type TaskStatus } from "../core/state.js";
import { applyTaskAnnotation, findTask, type Annotation } from "../core/tasksParser.js";
import { nextRunnable } from "../core/dag.js";
import { dagTasks } from "../core/render.js";
import { phaseForStatus } from "../core/phases.js";
import type { Args } from "../cli.js";

/**
 * Derive the execution-phase spec status from task states. SPEC defines no explicit phase-advance
 * command, and the agent may not edit state.json, so the CLI owns this transition: once execution
 * begins it reflects task progress. Planning statuses (requirements/design/tasks) are preserved
 * until at least one task leaves `pending`. When every task is complete the spec enters the VERIFY
 * phase (`verifying`) — a spec-level verification beat — and only `specd approve` finishes it to
 * `complete`/REFLECT, mirroring the per-task evidence gate at the spec level.
 */
function deriveStatus(state: ReturnType<typeof loadSpec>["state"]): void {
  const vals = Object.values(state.tasks);
  if (vals.length === 0) return;
  const started = vals.some((t) => t.status !== "pending");
  if (!started) return; // still in planning
  if (vals.every((t) => t.status === "complete")) {
    // Don't regress a spec a human already accepted (complete) back to verifying.
    state.status = state.status === "complete" ? "complete" : "verifying";
  } else {
    const next = nextRunnable(dagTasks(state));
    state.status = next.kind === "all-blocked" ? "blocked" : "executing";
  }
  state.phase = phaseForStatus(state.status);
}

const VALID = new Set<TaskStatus>(["complete", "blocked", "running", "pending"]);

export function run(args: Args): number {
  const root = requireSpecdRoot();
  const slug = args.pos[0];
  const id = args.pos[1];
  if (!slug || !id) throw usageError("usage: specd task <slug> <id> --status <complete|blocked|running|pending> [flags]");

  const status = args.flags.status;
  if (typeof status !== "string" || !VALID.has(status as TaskStatus)) {
    throw usageError("--status must be one of: complete, blocked, running, pending");
  }
  const newStatus = status as TaskStatus;

  // Hold the per-spec lock across load→mutate→dual-write so a parallel builder on another task of
  // the same spec can't clobber this update (§5.1). loadSpec re-enters the same lock harmlessly.
  return withSpecLock(root, slug, () => {
  const { state, doc } = loadSpec(root, slug);

  // Gate enforcement: don't mutate task status while a midreq approval is pending (override: --force).
  if (state.gate === "awaiting-approval" && args.flags.force !== true) {
    throw gateError(`spec '${slug}' is gated (awaiting-approval) — run \`specd approve ${slug}\` after the revised plan is approved, or pass --force`);
  }

  const ts = state.tasks[id];
  const docTask = findTask(doc, id);
  if (!ts || !docTask) throw notFoundError(`task '${id}' not found in spec '${slug}'`);

  const evidence = typeof args.flags.evidence === "string" ? args.flags.evidence.trim() : "";
  const reason = typeof args.flags.reason === "string" ? args.flags.reason.trim() : "";
  const stamp = nowIso();

  if (newStatus === "complete") {
    const incomplete = ts.depends.filter((d) => state.tasks[d]?.status !== "complete");
    if (incomplete.length) {
      throw gateError(`task ${id}: cannot complete — dependencies not complete: ${incomplete.join(", ")}`);
    }
    // G1 evidence gate: completion needs a deterministic verified record (specd ran the verify
    // line, exit 0) — NOT a free-text string on faith. The manual escape hatch (--unverified +
    // --evidence) covers read-only roles whose verify is N/A and genuinely-manual proofs.
    const unverified = args.flags.unverified === true;
    if (unverified) {
      if (!evidence) throw gateError(`task ${id}: --status complete --unverified requires non-empty --evidence`);
    } else {
      const verifyLine = (docTask.meta.verify ?? "").trim();
      const rec = ts.verification;
      if (!rec || !rec.verified) {
        throw gateError(`task ${id}: --status complete requires a passing \`specd verify ${slug} ${id}\` (exit 0) first — or pass --unverified with --evidence for a manual proof`);
      }
      if (rec.command !== verifyLine) {
        throw gateError(`task ${id}: verification is stale — recorded command (${rec.command}) ≠ current verify line (${verifyLine}); re-run \`specd verify ${slug} ${id}\``);
      }
    }
    ts.status = "complete";
    // Evidence string: the supplied proof, else a derived summary of the verified record.
    const derived = ts.verification && ts.verification.verified
      ? `verified: \`${ts.verification.command}\` → exit 0 @ ${ts.verification.gitHead ?? "no-git"} (${ts.verification.ranAt})`
      : "";
    ts.evidence = evidence || derived;
    ts.finishedAt = stamp;
    if (!ts.startedAt) ts.startedAt = stamp;
    ts.blocker = null;
    state.blockers = state.blockers.filter((b) => b.task !== id);
    docTask.checked = true;
    docTask.annotation = { kind: "complete", evidence: ts.evidence ?? "", ts: stamp };
  } else if (newStatus === "blocked") {
    if (!reason) throw gateError(`task ${id}: --status blocked requires --reason`);
    ts.status = "blocked";
    ts.blocker = reason;
    state.blockers = state.blockers.filter((b) => b.task !== id);
    state.blockers.push({ task: id, reason, since: `Turn ${state.turn}` });
    docTask.checked = false;
    docTask.annotation = { kind: "blocked", reason };
  } else if (newStatus === "running") {
    ts.status = "running";
    if (!ts.startedAt) ts.startedAt = stamp;
    ts.blocker = null;
    state.blockers = state.blockers.filter((b) => b.task !== id);
    docTask.checked = false;
    docTask.annotation = undefined;
  } else {
    // pending
    ts.status = "pending";
    ts.blocker = null;
    state.blockers = state.blockers.filter((b) => b.task !== id);
    docTask.checked = false;
    docTask.annotation = undefined;
  }

  // Dual write: tasks.md first (surgical line edit preserves comments/prose), then state.json
  // (source of truth) last. Each write is atomic.
  const tasksPath = artifactPath(root, slug, "tasks.md");
  const raw = readOrDefault(tasksPath, "");
  const updated = applyTaskAnnotation(raw, id, docTask.checked, docTask.annotation as Annotation | undefined);
  atomicWrite(tasksPath, updated);
  deriveStatus(state);
  saveState(root, slug, state);

  console.log(`task ${id} → ${newStatus}`);
  if (newStatus === "complete") console.log(`  evidence: ${ts.evidence}`);
  if (newStatus === "blocked") console.log(`  blocked: ${reason}`);
  return 0;
  });
}

File: reference/internal/cmd/task.go
package cmd

import (
	"fmt"
	"strconv"

	"github.com/0xkhdr/specd/internal/cli"
	"github.com/0xkhdr/specd/internal/core"
)

// ensureTelemetry lazily allocates a task's Telemetry record so callers can set
// a field without a nil check, keeping it omitted for tasks that never touch it.
func ensureTelemetry(ts *core.TaskState) *core.Telemetry {
	if ts.Telemetry == nil {
		ts.Telemetry = &core.Telemetry{}
	}
	return ts.Telemetry
}

// applyTelemetryAnnotations stores the operator-supplied --tokens/--cost values
// verbatim. specd records these as evidence; it never computes or prices them.
// Absent flags leave telemetry untouched (no empty record is created).
func applyTelemetryAnnotations(args cli.Args, ts *core.TaskState) {
	if args.Has("tokens") {
		if n, err := strconv.Atoi(args.Str("tokens")); err == nil {
			ensureTelemetry(ts).Tokens = n
		}
	}
	if args.Has("cost") {
		ensureTelemetry(ts).Cost = args.Str("cost")
	}
}

var validTaskStatuses = map[core.TaskStatus]bool{
	core.TaskComplete: true,
	core.TaskBlocked:  true,
	core.TaskRunning:  true,
	core.TaskPending:  true,
}

// RunTask implements `specd task`: it validates the requested --status
// transition for a task, then either delegates "complete" to
// core.CompleteTask's integrity-checked path or, under the spec lock,
// mutates the task's status/blocker/annotation directly, persisting both
// state.json and tasks.md.
//
//nolint:gocyclo // pre-existing complexity debt, out of scope for spec S3 — tracked for a future cleanup pass
func RunTask(args cli.Args) int {
	root, err := core.RequireSpecdRoot()
	if err != nil {
		return specdExit(err)
	}
	slug := ""
	id := ""
	if len(args.Pos) > 0 {
		slug = args.Pos[0]
	}
	if len(args.Pos) > 1 {
		id = args.Pos[1]
	}
	if slug == "" || id == "" {
		return usageExit("usage: specd task <slug> <id> --status <complete|blocked|running|pending> [flags]")
	}
	statusStr := args.Str("status")
	if !validTaskStatuses[core.TaskStatus(statusStr)] {
		return usageExit("--status must be one of: complete, blocked, running, pending")
	}
	newStatus := core.TaskStatus(statusStr)

	// Completion goes through the single core integrity path (the same one the
	// Pinky evidence reconciler uses), so dependency/verification/gate rules and
	// the state.json + tasks.md mutation live in exactly one place.
	if newStatus == core.TaskComplete {
		return runTaskComplete(root, slug, id, args)
	}

	rc, err := core.WithSpecLock[int](root, slug, func() (int, error) {
		loaded, err := core.LoadSpec(root, slug)
		if err != nil {
			return specdExit(err), err
		}
		state := loaded.State
		doc := loaded.Doc

		if state.Gate == core.GateAwaitingApproval && !args.Bool("force") {
			return specdExit(core.GateError(fmt.Sprintf("spec '%s' is gated (awaiting-approval) — run `specd approve %s` after the revised plan is approved, or pass --force", slug, slug))), nil
		}

		ts, hasTsState := state.Tasks[id]
		docTask := core.FindTask(doc, id)
		if !hasTsState || docTask == nil {
			return specdExit(core.NotFoundError(fmt.Sprintf("task '%s' not found in spec '%s'", id, slug))), nil
		}

		reason := args.Str("reason")

		switch newStatus {
		case core.TaskBlocked:
			if reason == "" {
				return specdExit(core.GateError(fmt.Sprintf("task %s: --status blocked requires --reason", id))), nil
			}
			ts.Status = core.TaskBlocked
			ts.Blocker = &reason
			core.AddBlocker(state, id, reason, state.Turn)
			docTask.Checked = false
			docTask.Annotation = &core.Annotation{Kind: core.AnnotBlocked, Reason: reason}

		case core.TaskRunning:
			ts.Status = core.TaskRunning
			if ts.StartedAt == nil {
				stamp := core.NowISO()
				ts.StartedAt = &stamp
			}
			ts.Blocker = nil
			core.RemoveBlocker(state, id)
			docTask.Checked = false
			docTask.Annotation = nil

		case core.TaskPending:
			ts.Status = core.TaskPending
			ts.Blocker = nil
			core.RemoveBlocker(state, id)
			docTask.Checked = false
			docTask.Annotation = nil
		}

		// Operator-annotated cost/usage (stored verbatim, never computed).
		applyTelemetryAnnotations(args, &ts)

		state.Tasks[id] = ts

		tasksPath := core.ArtifactPath(root, slug, "tasks.md")
		raw := core.ReadOrDefault(tasksPath, "")
		updated, err := core.ApplyTaskAnnotation(raw, id, docTask.Checked, docTask.Annotation)
		if err != nil {
			return specdExit(err), err
		}
		if err := core.AtomicWrite(tasksPath, updated); err != nil {
			return specdExit(err), err
		}
		cfg := core.LoadConfig(root)
		if routing, _, err := core.ResolveRoutingStamps(cfg.Routing, state.Tasks); err != nil {
			return specdExit(err), err
		} else if len(routing) > 0 {
			state.Routing = routing
		}
		core.DeriveSpecStatus(state)
		if err := core.SaveState(root, slug, state); err != nil {
			return specdExit(err), err
		}

		fmt.Printf("task %s → %s\n", id, newStatus)
		if newStatus == core.TaskBlocked {
			fmt.Printf("  blocked: %s\n", reason)
		}
		return core.ExitOK, nil
	})
	if err != nil {
		return specdExit(err)
	}
	return rc
}

// runTaskComplete drives the complete transition through core.CompleteTask and
// renders the same output the inline path used to. Telemetry flags are forwarded
// as verbatim annotations.
func runTaskComplete(root, slug, id string, args cli.Args) int {
	req := core.CompleteTaskRequest{
		Evidence:   args.Str("evidence"),
		Unverified: args.Bool("unverified"),
		Force:      args.Bool("force"),
	}
	if args.Has("tokens") {
		if n, err := strconv.Atoi(args.Str("tokens")); err == nil {
			req.Tokens = &n
		}
	}
	if args.Has("cost") {
		c := args.Str("cost")
		req.Cost = &c
	}

	res, err := core.CompleteTask(root, slug, id, req)
	if err != nil {
		return specdExit(err)
	}
	fmt.Printf("task %s → %s\n", id, core.TaskComplete)
	fmt.Printf("  evidence: %s\n", res.Evidence)
	return core.ExitOK
}
